# Advanced Python

Why Python? and if Python, then why advanced Python? Studying advanced Python yields multifaceted benefits, encompassing both personal and professional spheres. Advanced Python proficiency empowers individuals to engineer elegant and efficient solutions to complex problems, fostering innovation and driving organizational success.

1. **Enhanced Problem Solving:** Advanced Python proficiency equips practitioners with a diverse toolkit of advanced data structures, algorithms, and language features. This arsenal empowers developers to architect optimized and scalable solutions to intricate challenges, enhancing their problem-solving acumen.

2. **Code Efficiency:** Mastery of advanced Python constructs allows for the creation of concise and readable code. Proficient developers can optimize code execution, minimizing computational overhead and bolstering software performance.

3. **Elevated Productivity:** Familiarity with advanced Python libraries and frameworks expedites development cycles. Leveraging these tools enables the rapid creation of sophisticated applications, propelling developers towards heightened productivity.

4. **Career Advancement:** Proficiency in advanced Python is a valuable asset in the job market. Organizations seek adept Python developers for roles involving data science, machine learning, web development, and more. Mastery of advanced concepts widens career prospects and enhances earning potential.

5. **Innovation and Versatility:** Advanced Python enables the creation of cutting-edge applications in emerging domains such as artificial intelligence, machine learning, and data science. A strong foundation in advanced Python positions developers to drive innovation and contribute to transformative technologies.

6. **Collaborative Agility:** Adeptness in advanced Python fosters seamless collaboration among development teams. A common skill set streamlines communication, facilitates code reviews, and ensures consistent code quality.

7. **Software Maintainability:** Advanced Python practices emphasize modular design and encapsulation. This results in maintainable and extensible codebases, simplifying future updates and adaptations.

8. **Personal Growth:** Learning advanced Python cultivates cognitive skills like abstraction, algorithmic thinking, and pattern recognition. These skills transcend programming and contribute to holistic personal development.

9. **Contribution to Open Source:** Proficient Python developers can actively contribute to the open-source community. Sharing knowledge and collaborating on projects enriches the global programming ecosystem.

10. **Preparation for Emerging Trends:** As the technological landscape evolves, Python remains at the forefront. Adeptness in advanced Python ensures preparedness to embrace new paradigms, languages, and technologies that build upon this robust foundation.

In summation, the study of advanced Python offers a profound array of advantages, encompassing technical prowess, career progression, and personal growth. By delving into its intricacies, individuals position themselves as agile and influential contributors to the dynamic world of software development.

## Part I: Object-Oriented Programming

### Classes

Object-Oriented Programming (OOP) in Python entails the creation of blueprints, facilitating structured software development. A class, denoted with an uppercase name, serves as this blueprint and encapsulates functions, referred to as methods. A class is typically populated with methods, yet if desired, a `pass` command can render it inactive. This concise definition encapsulates the essence of OOP principles.

In [8]:
class Dog:
    pass

The following shows how one can add variables/attributes in the class. These are global private attributes within the class. It can only be accessed in the class.

In [9]:
class Dog:
    info = "4 legged canine mammal." # <- PRIVATE attribute holds a string
print(info) # The print command is outdented from the class where 'info' does not exist

NameError: name 'info' is not defined

The `NameError` has been raised, as `info` does not exist outside the class `Dog`. It is a private attribute of the class.

In order to print `info` we need to call the class and then the associated variable.

Like this:

In [ ]:
print(Dog.info)

4 legged canine mammal.


Another example of a class (`MacBook_Pro`) with an attribute inside it. The attribute is then called outside the class, using the appropriate syntax.

In [ ]:
class MacBook_Pro:
    '''
    Describes Macbook_Pro with a select few attributes.
    '''
    year = 2019 # type: int
    material = "Aluminium" # type: str
    company = "Apple" # type: str
print(f"""  The following are attributes, extracted from the class
        The year the laptop model was released: {MacBook_Pro.year}
        The material the laptop was made with: {MacBook_Pro.material}
        The company that made it: {MacBook_Pro.company}""")

  The following are attributes, extracted from the class
        The year the laptop model was released: 2019
        The material the laptop was made with: Aluminium
        The company that made it: Apple


#### Instances

As per the name suggests, instances are objects that are created when the class is called out in a command in the code.

 The term instance and object are interchangeable. The following is how the `Dog` class is called as an instance.

In [ ]:
Dog() # this is an instance/object of CLASS: Dog

Here the `Dog` class was called out.

Now, every time the `Dog` function is called out nothing is initialised. To make this class usable, we initialise a function within the class `Dog`, henceforth every time the `Dog` is called, the function along with will get initialised.

This function is all also called the `__init__` function. Whenever an object is created from this class, this function is initiated.

In [33]:
class Dog:
    print("ClASS: Dog has been declared")
    def __init__(self): # initialsation function made within the class 'Dog'
        print("Woof Woof mf")

ClASS: Dog has been declared


In [34]:
Dog()

Woof Woof mf


Each and every instance (call-out) of the class is unique. They all are individual objects created from one single class `Dog`.

Variables within the class can be created so that it can be called out when required by an instance. An instance is unique (as mentioned) and by creating a variable in the self function, we can utilise it provide a variable to each unique and individual object/instance.

In [17]:
import random # importing random to test whether all the objects from a class are unique

class Dog:
    def __init__(self):
        self.testNum = random.randint(1,10)

In [19]:
dog1 = Dog() # one instance/object of 'Dog'
dog2 = Dog() # another instance/object of 'Dog'

print(f"""First instance/object using the variable {dog1.testNum}
Second instance/object using the variable {dog2.testNum}""")

First instance/object using the variable 8
Second instance/object using the variable 8


Here, what exactly happened was that there were 2 instances called from `Dog`, `dog1` and `dog2`, these two objects utilised the `__init__` function with the `self` command. It can be seen that each object is unique as the random numbers created by using the `__init__` function is unique.

*NOTE: `dog1` and `dog2` are unique objects/instances.*

In order to create a variable from the class blueprint, the `__init__` function can be modified. A set of empty variables can be established in order for the object to use and fill them up.

For example, we have a blueprint `Dog`, from there we can include a set of commands in the `__init__`. It provides functions to provide the facilities for the object to take in the `name` and `breed` of the dog. That attribute of the object (`Dog()` instances) will be unique to the dog itself. I.e., that dog will carry its own value for `name` and `breed`, but it will have the ability due to the function in class `Dog`.

In [28]:
class Dog: # Class
    def __init__(self,name,breed): # The __init__ function holds itself (self) and the name and breed initiation for the variables 
        self.name = name # initiating the name variable
        self.breed = breed # initiating the breed variable
         
# utilising the variable capacity from the __init__ function:
dog_first = Dog('Doggie','The brown kind') # the variables are in order of the the establishment as per the __init__ function
dog_second = Dog('Kutta','White wala')     # i.e., the name comes first and then breed

print(f"""The name of the first dog is {dog_first.name} and the breed is {dog_first.breed}.
Whereas, the name of the second dog is {dog_second.name} and the breed is {dog_second.breed}.
      """)

The name of the first dog is Doggie and the breed is The brown kind.
Whereas, the name of the second dog is Kutta and the breed is White wala.
      


Here, the class `Dog` contained a `__init__` function which helped create empty variables for `name` and `breed`. Every time a new instance of `Dog` was made, the values for the instance based on the `__init__` function had to be made. They need to be made in that exact order too!

In the `__init__` function we have initialised `name` and `breed`, which are both in order `name, breed`. This order has been followed when establishing an object/instance of class `Dog`, namely `dog_first` and `dog_second`.

#### Methods

Simply, a method is function inside a class (`def`). The following shows a class `Square`.

Here we have provided an initialisation function with `__init__`, and it holds the main variable for any instance that follows; i.e., `side`.

We then use the `side` variable to a method which is named `area` (it's literally a function in a class). In this method the `side` variable is called, and it is used to find the `area` of the square.

In [4]:
class Square: # main class
    def __init__(self, side): # __init__ function
        self.side = side # intitialised "side" variable for any instances
    def area(self): # the actual method from CLASS: Square to find area of a square
        return (self.side **2) # area of a square
sqaure_side = Square(3) # value for the length of the side [instance]
area1 = Square.area(sqaure_side) # the area of the sqaure [instance of "def area"]

print(f"the side is {sqaure_side.side} and its area is {area1}")

the side is 3 and its area is 9


Here the `sqauare_side` instance helps declare a value for the length of the side of a said-given square. Whereas, the `area1` instance utilises the `square_side` object's value to find area of the square.

It is imperative that when a variable is used within a method, it has to be called using the `self.` function, otherwise it will not be recognized. It does not always have to be `self` in particular. We use `self` as it is convention.

#### Inheritance

Consider the following code block:

In [5]:
class Phone: # master class
    def __init__(self):
        print("Phone class is created")

class iPhone(Phone): # inherited class from 'master class'
    def __init__(self):
        super().__init__(self) # activates the parent class
        print('Iphone is created')
        
class iPhone_12pro(iPhone): # inherited class from 'iPhone'
    def __init__(self):
        super().__init__(self) # activates the parent class
        print("Iphone 12 Pro is created")

Master Class: Imagine you're making a bunch of different phones in a phone factory. To make this easier, you decide to use a **master blueprint** for all phones. This blueprint is called the `Phone` class. It has some basic instructions for creating a phone.

Inherited from Master Class: Now, you want to make a special type of phone that's like a regular phone but with some extra features. Let's call it `iPhone`. To do this, you create a new blueprint called `iPhone` that is based on the `Phone` blueprint. This is called "inheritance."

Inherited even further: Great! Now you have regular phones and iPhones. But wait, you want an even fancier phone – the `iPhone_12pro`. It's like an `iPhone`, but with even more special features. So, you create another blueprint, `iPhone_12pro`, based on the `iPhone` blueprint.

Here's what happens when you create an `iPhone_12pro`:

1. The `iPhone_12pro` blueprint says, "Hey, I'm based on an `iPhone`." So, it goes to the `iPhone` blueprint.

2. The `iPhone` blueprint says, "Hey, I'm based on a `Phone`." So, it goes to the `Phone` blueprint.

3. The `Phone` blueprint says, "Alright, let's follow my instructions." It prints "Phone class is created."

4. Then, the `iPhone` blueprint continues its instructions. It prints "iPhone is created."

5. Finally, the `iPhone_12pro` blueprint finishes its instructions. It prints "iPhone 12 Pro is created."

This way, you can create different types of phones with shared characteristics, and each new type can add its own special features while reusing the instructions from the parent blueprints.

### Try and Except

*Try&Except* is a debugging tool to pinpoint certain error so that if in any case the said error occurs: it does not crash our program. This is on the basis that we anticipate the code and know what can/should go wrong and request python to tackle it.

Python simply tries to fulfil our request and if the request is impossible to fulfil it raises our exception token and moves on. If it would not do that, then the code for absolutely come to a stop there and would not proceed. This is great for debugging.

Python is line-by-line executed language and if a previous line is not run, or it faces an error it will absolutely stop there and not do anything further. *Try&Except* helps us not stop due to an error and gives us a warning about an error and moves on.

The following encapsulates this idea:

In [9]:
print("Before Try&Except")

try:
    2/0 # this should raise an error
except:
    print("MATH ERROR: Can't divide with 0! but we roll") # this should be the raised error token

print("After the Try&Except") # this line should be executed due to Try&Except

Before Try&Except
MATH ERROR: Can't divide with 0! but we roll
After the Try&Except


If *Try&Except* would not be used:

In [10]:
print("Before Try&Except")

2/0 # this will raise a ZeroDivisionError and nothing here onwards will be executed

print("After the Try&Except")

Before Try&Except


ZeroDivisionError: division by zero

The compiler has raised a `ZeroDivisionError: division by zero` reply in the console. This simply due to the fact that mathematically nothing can be divided by 0. But the real point here is to notice how we are stopped/stuck at the `ZeroDivisionError` and can't move forward to print the statement in line `5`.

We can make our code even more intuitive and reciprocate to our errors by the type they are. For example, we can raise a custom `NameError` when a non-declared variable is called.

In [45]:
try:
    print(test) # 'test' is not defined
except NameError:
    print("This was a NameError as an undefined variable has been called!") # hence, this particular error is raised

This was a NameError as an undefined variable has been called!


So now our `except` statement is even more detailed and can identify the type of error that has occurred and raise the particular `except` message. This `NameError` is issued only when a `NameError` takes place, and not when any other general error takes place.

We can also customise our types of error for project development. These types of errors are bespoke and are raised only when those particular errors are raised.

For example, let's say we are building a calculator and our calculator cannot take string as input data. We need to raise an error that informs the user that string is impossible to be input into the calculator.

There are steps to this, and they are as follows.

1. #####  Create a verification class `InputError`

In [32]:
class InputError(Exception): # this empty error class is inherited from the 'Exception' function
    pass

`InputError` is our special/unique error token. 

2. ##### Create a function to identify and activate the `InputError` (this could be any function you wish, it just needs to make use of the *Try&Except* feature for now)

The following function `upper_convertor`, takes a string input and converts all its characters to upper-case. If the user has not supplied any input, i.e., has given an empty string as input: the custom error will be raised

In [2]:
def upper_convertor(input: str):
    """
    Takes in a string input and converts all of the string's characters into upper-case characters

    Args:
        input (str): A string input for the convertor
    
    Returns:
        str: The converted output from lower to uppercase
    """
    if len(input) <= 0: #checks if input is empty by seeing its size: if size == 0 then 'empty'
        raise InputError("Empty input!") # raises the custom error
    return input.upper()

3. ##### Testing the error

In [44]:
print(upper_convertor('')) # tests the custom error by giving it an empty string

InputError: Empty input!

The error raised is our own custom `InputError` in red.

The following is a run with compatible input.

In [3]:
print(upper_convertor("abcd"))

ABCD


## Part II: Project - *Weather API*

Here we are going to build a Weather Application using an API from [OpenWeather API](https://openweathermap.org/api). 

The API key is: `70ff51bbc74bdcd001ca0dce3739b09f`


Using the latitude and longitude value for a location, we can extract the weather data for that place. The output data is a *JSON* file. A *JSON* is a data output type for API's. It essentially is a huge `dictionary`.

From the API documentation we can get a call function which is: https://api.openweathermap.org/data/2.5/weather?lat={lat}&lon={lon}&appid={API_key}. This API function can have the latitude and longitude with the API key to get information for any given location. For example, the current weather for Dubai, UAE can we called by giving 25.2048° N, 55.2708° E for `lat` and `long`.

The following is the output:

`{"coord":{"lon":55.2708,"lat":25.2048},"weather":[{"id":800,"main":"Clear","description":"clear sky","icon":"01n"}],"base":"stations","main":{"temp":307.17,"feels_like":314.17,"temp_min":303.36,"temp_max":307.33,"pressure":996,"humidity":70},"visibility":10000,"wind":{"speed":3.09,"deg":150},"clouds":{"all":0},"dt":1687816994,"sys":{"type":1,"id":7537,"country":"AE","sunrise":1687829448,"sunset":1687878763},"timezone":14400,"id":290845,"name":"Ash Shindaghah","cod":200}`

### `Request` library/package

We call for the `requests-2.31.0` package/library in python so that we can call and mess around with our *JSON* file from the API's web response.

In [2]:
import requests

In the following code block I have the latitude and longitude value for Dubai, UAE. Along with that I have my unique API key.

In [24]:
lat_ = 25.2048
lon_ = 55.2708
key = '70ff51bbc74bdcd001ca0dce3739b09f'

The following code block uses the `requests` library to get the API response and manipulate to use in development. A *Try&Except* has been used to make sure that if the user is not connected to the internet, an error token explicitly addressing that is raised.

In [25]:
try:
    response = requests.get(f"https://api.openweathermap.org/data/2.5/weather?lat={lat_}&lon={lon_}8&appid={key}") # API request
    print(response.json()) # Print the request
except ConnectionError as e:
    print("No internet connection!") # raise token if connection error occurs

{'coord': {'lon': 55.2708, 'lat': 25.2049}, 'weather': [{'id': 800, 'main': 'Clear', 'description': 'clear sky', 'icon': '01n'}], 'base': 'stations', 'main': {'temp': 307.17, 'feels_like': 314.17, 'temp_min': 303.36, 'temp_max': 307.33, 'pressure': 996, 'humidity': 66}, 'visibility': 10000, 'wind': {'speed': 4.12, 'deg': 110}, 'clouds': {'all': 0}, 'dt': 1687822754, 'sys': {'type': 1, 'id': 7537, 'country': 'AE', 'sunrise': 1687829448, 'sunset': 1687878763}, 'timezone': 14400, 'id': 290845, 'name': 'Ash Shindaghah', 'cod': 200}


### Working with the *JSON* data

In order to work with the API and its *JSON* response, a variable `response_json` has been created. This variable holds the entire *JSON* response from the API and can be manipulated to provide a more concise format of the output.

In [26]:
response_json = response.json() # assigns a variable through the JSON file with API response

In [27]:
response_json

{'coord': {'lon': 55.2708, 'lat': 25.2049},
 'weather': [{'id': 800,
   'main': 'Clear',
   'description': 'clear sky',
   'icon': '01n'}],
 'base': 'stations',
 'main': {'temp': 307.17,
  'feels_like': 314.17,
  'temp_min': 303.36,
  'temp_max': 307.33,
  'pressure': 996,
  'humidity': 66},
 'visibility': 10000,
 'wind': {'speed': 4.12, 'deg': 110},
 'clouds': {'all': 0},
 'dt': 1687822754,
 'sys': {'type': 1,
  'id': 7537,
  'country': 'AE',
  'sunrise': 1687829448,
  'sunset': 1687878763},
 'timezone': 14400,
 'id': 290845,
 'name': 'Ash Shindaghah',
 'cod': 200}

To pick particular bit of information from the *JSON* file (instead of the whole nested dictionary), we look for the key and the main key for that particular information.

For example, to access the `main` climate information from the master nested dictionary: we need to call `main` from the master dictionary like so. 

In [28]:
response_json['main']

{'temp': 307.17,
 'feels_like': 314.17,
 'temp_min': 303.36,
 'temp_max': 307.33,
 'pressure': 996,
 'humidity': 66}

To get further particular with the nested dictionary, we simply add the desired key after the respective master key. In order to call the temperature we use the `temp` key from `main`.

In [29]:
response_json['main']['temp']

307.17

We see how this value for the temperature is absurd. That is because the API uses `standard` meteorological units. We can change this by modifying the API call function URL. 

As per the documentation:
*Units of measurement: Standard, metric and imperial units are available. If you do not use the units' parameter, standard units will be applied by default.*

So we can add `metric` as a unit in the URL. This makes the new URL the following:

https://api.openweathermap.org/data/2.5/weather?metric&lat={lat}&lon={lon}&appid={API_key}

NOTE: The separator here is `<b>&</b>`, and it is boolean for `and`, i.e., the API calls for `unit` & `lat` & `lon` etc.

In [68]:
lat_ = 25.2048
lon_ = 55.2708
Place = 'Dubai, UAE'
key = '70ff51bbc74bdcd001ca0dce3739b09f'
response = requests.get(f'https://api.openweathermap.org/data/2.5/weather?units=metric&lat={lat_}&lon={lon_}&appid={key}')

Now if we get the temperature, the units should be in metric system.

In [69]:
response_json_new = response.json()
response_json_new

{'coord': {'lon': 55.2708, 'lat': 25.2048},
 'weather': [{'id': 800,
   'main': 'Clear',
   'description': 'clear sky',
   'icon': '01d'}],
 'base': 'stations',
 'main': {'temp': 38.27,
  'feels_like': 45.27,
  'temp_min': 37.21,
  'temp_max': 41.18,
  'pressure': 995,
  'humidity': 60},
 'visibility': 7000,
 'wind': {'speed': 7.2, 'deg': 320},
 'clouds': {'all': 1},
 'dt': 1687876638,
 'sys': {'type': 1,
  'id': 7537,
  'country': 'AE',
  'sunrise': 1687829448,
  'sunset': 1687878763},
 'timezone': 14400,
 'id': 290845,
 'name': 'Ash Shindaghah',
 'cod': 200}

In [70]:
response_json_new['main']['temp']

38.27

We can assign these API responses to different kinds of apt variables, this way can utilise the *JSON* data with those variables.

In [71]:
current_temp = response_json_new['main']['temp']
temp_min = response_json_new['main']['temp_min']
temp_max = response_json_new['main']['temp_max']

For example, we can use the variables with API data as:

In [74]:
print(f'The current temp at {lat_}º N {lon_}º E {Place} is {current_temp}º C. The low will be {temp_min}º C, while the high will be {temp_max}º C.')

The current temp at 25.2048º N 55.2708º E Dubai, UAE is 38.27º C. The low will be 37.21º C, while the high will be 41.18º C.


### Creating the final modular system to use API for data presentation

We instantiate a class called `City` which hold all the methods/functions to interact with the user and the API. It can use that information to present the data for any given set of co-ordinates (latitudes and longitudes).

In [125]:
class City:

    def __init__(self, Place, lat, lon,units='metric'): # __init__ to initiate variables
        self.Place = str(Place)     # as the parameter suggests: Place name variable for the class (str)
        self.lat = float(lat)       # as the parameter suggests: latidtude variable for the class (float)
        self.lon = float(lon)       # as the parameter suggests: longitude variable for the class (float)
        self.units = str(units)     # as the parameter suggests: temp unit variable for the class (str)
        self.get_json()  # external function initiated

    def get_json(self):
        """
        Takes in the response from the OpenWeather API and stores it in a variable called `response_API`
        Args:
            self: initialiser
            response_API: (JSON) API's response
            response_city: (JSON) Response data stored into variable for use
        Returns: 
            current_temp: (JSON) Current temperature
            temp_min: (JSON) Minimum Temperature
            temp_max: (JSON) Maximum Temperature
        """
        try: # tries to call for API response 
            response_API = requests.get(f'https://api.openweathermap.org/data/2.5/weather?units={self.units}&lat={self.lat}&lon={self.lon}&appid=70ff51bbc74bdcd001ca0dce3739b09f')
        except ConnectionError as e: # if there is a connection error then it raises that token
             print("No internet connection!")

        self.response_city = response_API.json() # stores the response in an class-local variable

        self.current_temp = self.response_city['main']['temp'] # stores current temperature
        self.temp_min = self.response_city['main']['temp_min'] # stores low temperature
        self.temp_max = self.response_city['main']['temp_max'] # stores high temperature

    def get_weather(self): 
        """
        Takes in the response from the user on their requested location with desired unit
        Args:
            self: initialiser
            self.: all __init__ variables declared in __init__ function
            units_symbol: default celsuis decleration
        Returns: 
            f-string (str): the requested weather indicators for the location requested in desired unit
        """
        units_symbol = "C" # default unit of temperature as per the __init__ function
        if self.units == "imperial": # if statement when farenheit is requested 
            units_symbol = "F" # farenheit symbol
        print(f'''The current temp in {self.Place} (co-ords: {self.lat}º, {self.lon}º) is: {self.current_temp}º {units_symbol}.
The low will be {self.temp_min}º {units_symbol}, while the high will be {self.temp_max}º {units_symbol}.''') # the final ouput f-string with all requested information

<b>*It is essential that the prompt information to get the weather is in exact order as the `__init__` function, as the compiler reads by corresponding the prompt to the function.*</b>

Example with London, UK:

In [118]:
london_weather = City('London',51.5072,0.1276)
london_weather.get_weather()

The current temp in London (co-ords: 51.5072º, 0.1276º) is: 19.38º C. 
The low will be 17.53º C, while the high will be 20.15º C.


Example with New Delhi, India:

In [119]:
NewDelhi = City('New Delhi',28.644800,77.216721)
NewDelhi.get_weather()

The current temp in New Delhi (co-ords: 28.6448º, 77.216721º) is: 28.07º C. 
The low will be 28.07º C, while the high will be 28.07º C.


In [127]:
NewYork = City('New York', 40.730610,-73.935242,units='imperial')
NewYork.get_weather()

The current temp in New York (co-ords: 40.73061º, -73.935242º) is: 73.26º F.
The low will be 68.29º F, while the high will be 77.36º F.


## Part III: Functional Programming Concepts

### Advanced comparisons and `If` statements

When comparing multiple parameters to assess a given condition, using boolean algebra helps significantly. It uses concepts of multi-variate/bi-variate parameter comparison on a binary level. A given `if` statement can be tuned to look out for pairs/groups that satisfy (or dissatisfy) the boolean conditions such as `and`, `or`, `nor`, etc.

This attribute helps look out for code that needs to be executed only when a particular set of criteria has been met.

For example, if a company is hiring new employees it will look for certain qualifications or a set of qualifications. Those qualifications can utilise a combination or a one-or-other system. 

<b>Condition statement</b>: Let us consider that an applicant will get hired if they have an undergraduate degree *or* if they have a valid diploma. Whilst possessing one-or-the-other, they <b>must</b> have scored 65 or above in their school. This grade is meant to be an aggregate value.

In [15]:
UnderGrad = True
Diploma = False
SchoolGrades_agg = 64
LetterOfRec = True

if (UnderGrad or Diploma) and SchoolGrades_agg >= 65: # checks if person has UnderGrad or Diploma, and 65 in school grades
    print("You are hired!") # yes they do
elif LetterOfRec: # read as 'else-if' 
    print("You are hired!") # the don't match one of the criteria but they have an LOR so it works
else:
    print('You are not hired.') # none of the conditions are met

You are hired!


Some boolean functions can be simplified by using an equally (if not more) effective notation.

Here's a table of the equivalent operators in both formats:
<center>

|  Boolean |  Python |
|:--------:|:-------:|
|   `and`  |   `&`   |
|   `or`   |   `\|`  |
|   `not`  |   `~`   |
|   `xor`  |   `^`   |

</center>

In [1]:
UnderGrad = True
Diploma = False
SchoolGrades_agg = 64
LetterOfRec = True

if (UnderGrad | Diploma) & SchoolGrades_agg >= 65: # checks if person has UnderGrad or Diploma, and 65 or more in school grades
    print("You are hired!") # yes they do
elif LetterOfRec: # read as 'else-if'
    print("You are hired!") # they don't match any of the criteria but they have an LOR so it works
else:
    print('You are not hired.') # none of the conditions are met

You are hired!


### Switch (`match`) Statements

As the name suggests this function switches/matches one data unit with another. This is referred as a `switch` statement in the programming industry and this can be used to match a given data entry with another. For example, in the code snippet below, the user's input numbers are *switched* with a word-based representation of it. 

Each number has its curated case, and that case helps match/switch the data entries.

It is important to note that the user can only input numbers between 1 and 5. If the input is out of the range then it implies then the `case _` keyword helps raise a `exception` token.

In [5]:
number = input("Enter number between 1 and 5: ") # user input is stored in number 

match number: # switch statement
    case 1: # SYNTAX: "case <data entry>:"
        print('One') # SYNTAX: "print(<corresponding data entry>)"
    case 2:
        print('Two')
    case 3:
        print('Three')
    case 4:
        print('Four')
    case 5:
        print('Five')
    case _: # unknown response
        print('invalid response')

invalid response


### Lambda (`lambda`)

<b>*This is simply a one-line function for functional programming concepts to utilise. It does what `def()` would do, except in one line for the functional programming logic.*</b>

### Maps

Maps help reduce lines of code and makes the programming logic more efficient. It utilises a `lambda` initiator to create a sub-function which uses standard python methodologies within.

In [7]:
class Student:
    """
    DocString:
    This is a class for students with their names and scores in different subjects/courses of study.
    """
    def __init__ (self, name, score):
        """
        DocString:
        Initializes the student object by setting its attributes to given values or default ones if not provided.
        Args:
            self: initialisor 
            name: str 
            score: float
        returns: null
        """
        self.name = name
        self.score = score

students = [Student("Joe", 0.46), Student("Amy", 0.72), Student("Mark", 0.88), Student("Zach", 0.75), Student("Jane", 0.84), Student("Sarah", 0.63)] # score data as a list

In [11]:
for student in students:
    print(type(student.score))

<class 'float'>
<class 'float'>
<class 'float'>
<class 'float'>
<class 'float'>
<class 'float'>


In [8]:
map_results = list(map(lambda student: f"{student.name} Passed" if student.score >= 0.7 else f"{student.name} Failed", students)) # map with lambda and if-else statement
map_results

['Joe Failed',
 'Amy Passed',
 'Mark Passed',
 'Zach Passed',
 'Jane Passed',
 'Sarah Failed']

In [9]:
numbers = [1,2,3,4,5]

print(list(map(lambda number: number * 2, numbers)))

[2, 4, 6, 8, 10]


### Filter

As the name suggests the `filter()` filters out data elements based on the given conditions. We use `lambda` here again to initiate the element from a list. We then use that element to set the condition.

The filtered data can be stored in a variable as a new list then.

In [15]:
filter_number = list(filter(lambda number: number%2 == 0,numbers)) # filters out all even numbers and presents them
filter_number

[2, 4]

### Reduce

This can help reduce or deduce a logical aspect from a list or data. It uses the `lambda` to initiate a function and that function is used to perform specified logic on the entire dataset (entire list).

In [20]:
from functools import reduce
# numbers = [1,2,3,4,5] has already been declared before
reduce_mutiple = reduce(lambda total, number: number*total, numbers,  1) # multiplies everything in a dataset
reduce_mutiple

120

## Part IV: Project - *Web Scraper*

In this section we will create a project that can utilise information off of a website, especially in a condition where there is no explicit *API* available.

The following code block simply returns the `URL`'s response to us. This response is **not** the content of the webpage mentioned in `URL`.

In [5]:
import requests # 'request' library

URL = 'https://linkedin.com' # URL for 'LinkedIn.com'
response = requests.get(URL) # stores the URL's response in a 'response' variable

response # outputs 'response' varible's information

<Response [200]>

The output `<Response [200]>` in *HTTPS* terms, this implies "OK"; as in the return of the request has been successfully received. The following code block does exactly same.

In [4]:
print(response.status_code)

200


To get the content from this webpage we can use the `.content` method, and we will get the page's source-code in return. This has not been shown below as it is a massive block of code and would take too much space causing congestion and eye-stain.

In [ ]:
response.content # returns the page's source-code

### *Pixelford*

We now want to call a website called *Pixelford*. *Pixelford* is a fictitious website created simply to mess around with. It is an arbitrary photography consultancy website that hosts blogs, gig requests, and about pages. We are going to use the `requests` again to get an "OK" (`<Response [200]>`) response from the *URL* making sure everything is in order, i.e., we are communicating with the page as intended.

*Pixelford*'s blog page URL: https://pixelford.com/blog/

In [6]:
pixelford_URL = 'https://pixelford.com/blog/' # URL for 'PIXELFORD'
response = requests.get(pixelford_URL) # stores the URL's response in a 'response' variable

response # outputs 'response' varible's information

<Response [403]>

It can be seen that our response from *Pixeford* returns a response type `403`. This implies that the website's server is forbidden to us. This occurrence provides valuable insights into web scraping challenges. The website we're scraping, *Pixelford*, is an example website from *LinkedIn Learning*, used for learning purposes. However, due to high student requests post-course release, the website's server software flagged excessive requests from tools like Python as potentially malicious, causing the blocking of such requests. This is due to the **user agent** information transmitted with web requests, indicating the device and browser being used. Browsers usually provide detailed user agent data, while Python's requests may have a different user agent. This discrepancy leads to the server perceiving Python requests as suspicious.

The only practical way to work around is to change the **user agent**'s header data. Changing the **user agent**'s header data will make the computer believe that not any random python user is requesting the information, but a particular user is, which it will not consider malicious (or forbid-worthy). By that it is implied the **user agent**'s token be changed. This can be done by adding the `requests` library's `headers = {'user-agent': '<whatever unique/random name we want for our user>'}` assignment operator to the `.get` request. This has been shown below.

In [8]:
pixelford_URL = 'https://pixelford.com/blog/' # URL for 'PIXELFORD-blogs'
response_without403 = requests.get(pixelford_URL, headers = {'user-agent': 'pineapple'}) # using the 'headers' assignment operator to request using a unique alias

response_without403 # outputs 'response_without403' varible's information

<Response [200]>

We now get `<Response [200]>` which is a sign from *HTTPS* this is running "OK".

### Beautiful Soup

When we get requested content from webpages, it comes directly from their source code. This output is usually not user-friendly to look at specifically, to counter that we utilise *Beautiful Soup*.

In [9]:
from bs4 import BeautifulSoup # imports the required dependency

Our aim: to webscrape all the titles from the "blog" webpage in *Pixelford*. When the source-code of this page is inspected, we come to realise that all the titles are under *a-tags* (`<a>...</a>`). So we call all the *a-tagged* content from the page. For that we first need to parse the received large content.

In [10]:
html = response_without403.content # calls all the content from our previously requested URL
soup = BeautifulSoup(html, 'html.parser') # parses the large (ugly) content into digestable sets of data

Now we can find all *a-tags* from our webscraped and parsed content. This can be done by using the `final_all()` function of *Beautiful Soup*. We use parameter `'a` and call all instances of *a-tags*.

In [11]:
a_tags = soup.find_all('a') # finds all instances of a_tags in the tag headers and stores in 'a_tags'
a_tags # returns all data in 'a_tags'

[<a href="https://pixelford.com/">Pixelford Photography</a>,
 <a href="https://pixelford.com/about/" itemprop="url"><span itemprop="name">About</span></a>,
 <a aria-current="page" href="https://pixelford.com/blog/" itemprop="url"><span itemprop="name">Blog</span></a>,
 <a href="https://pixelford.com/services/" itemprop="url"><span itemprop="name">Services</span></a>,
 <a href="https://pixelford.com/process/" itemprop="url"><span itemprop="name">Our Process</span></a>,
 <a href="https://pixelford.com/pricing/" itemprop="url"><span itemprop="name">Pricing</span></a>,
 <a href="https://pixelford.com/portfolio/" itemprop="url"><span itemprop="name">Portfolio</span></a>,
 <a href="https://pixelford.com/baby/" itemprop="url"><span itemprop="name">Babies</span></a>,
 <a href="https://pixelford.com/engagement/" itemprop="url"><span itemprop="name">Engagement</span></a>,
 <a href="https://pixelford.com/family/" itemprop="url"><span itemprop="name">Family</span></a>,
 <a href="https://pixelford.

As it is visible, we have received all the data where *a-tags* exist. But it can be seen that we have received every single instance of an *a_tag*, including the links and link-pages, we don't need it. So, we need to request in order to get an even cleaner looking output. The aim is to get an output with **only** the titles. We can do that by establishing a `class_` parameter and setting it to `'entry-title-link'`. We got the `'entry-title-link'` parameter from the source-code, it is subjective to vary.

In [13]:
a_tags_title = soup.find_all('a', class_ = 'entry-title-link')

for a_tag in a_tags_title:
    print(a_tag.get_text()) # periodically prints all the titles as text outputs

Get Rid of Photo Bombers
The External Light Meter: Your New Best Friend
Bridal Gown: Capturing the Details of the Dress
Shooting a Wedding
The Growth of Baby Jill
Jake Neopolt, “Fire in the Sky” album cover
Chen Family
Juliane and Hanz
The Authentic Charm of Coney Island
Brooklyn Bridge – One Bridge, So Many Options


Now that we have our titles retrieved, let us save it as a data-pack.

In [14]:
titles = list(map(lambda a_tag: a_tag.get_text(), a_tags_title)) # maps all the titles, converts into texts, and saves them as a list into var: 'title'

With this all titles are saved into list called `titles`. This can be used to manipulate the data as use it for $n$ number of purposes.

### Final Webscraper

Our current web scraper is quite impressive. It diligently retrieves blog post titles and displays them in a neat list. However, the blog posts contain more than just titles. Notably, the publication date holds our particular interest. The idea is to exhibit the post's creation date, enabling users to spot new content effortlessly.

Upon examination, we notice the date lies just below the *a-tag*, which holds the blog post title. It resides within a *time* tag, offering precise details: the year, month, day, hour, minute, and second. While a well-formatted "September 13th, 2019" is tempting, we can also extract the detailed `datetime` if necessary.

The plan is to revamp our Beautiful Soup code. Instead of targeting *a-tags*, we'll focus on articles (blogs). This approach aligns with our reference to "blog posts." By searching for the *article* tag, we filter content effectively. Our code now captures all the articles, and to fine-tune, we look for the class `type post` within the article.

Next, we zero in on the title and date within each blog. To access the title, we employ Beautiful Soup's `find` method on the blog element, looking for an *a-tag* with class `entry title link`. The title is extracted and simplified for clarity.

Parsing the date proves equally insightful. The date lies within a *time* tag labeled `entry time`. We retrieve this tag and subsequently transform the `datetime` string into a more manageable `datetime` object using Python's built-in `datetime` module.

Our `datetime` object now takes center stage. We desire a personalized date representation, and Python's `strftime` method comes to the rescue. This method lets us format the `datetime` as desired. We visit `strftime.org` to comprehend the syntax for customizing the date presentation.

Finally, with both the title and the prettified date at hand, we use an F-string to combine them into a cohesive, user-friendly format. This presents a streamlined list of blog posts, complete with dates and titles.

Keep in mind that web scraping's effectiveness hinges on website consistency. Should the site structure alter, our scraper could falter. However, for targeted use cases, web scraping proves an engaging method to fetch information dynamically.

In [16]:
# Import the datetime module to work with date and time objects
import datetime

# Find all the article tags with the class "type-post"
blogs = soup.find_all('article', class_="type-post")

# Loop through each blog post
for blog in blogs:
    # Find the anchor tag (a-tag) with the class "entry-title-link" and get its text (title)
    title = blog.find('a', class_="entry-title-link").get_text()

    # Find the time tag with the class "entry-time" and get the value of its 'datetime' attribute
    blog_datetime_string = blog.find('time', class_="entry-time").get('datetime')

    # Convert the datetime string to a datetime object using fromisoformat
    blog_datetime = datetime.datetime.fromisoformat(blog_datetime_string)

    # Format the datetime object as a pretty date (e.g., "Sep 01 2023")
    pretty_date = blog_datetime.strftime("%b %d %Y")

    # Print the formatted date and the title of the blog post
    print(f"{pretty_date} - {title}")

Sep 13 2019 - Get Rid of Photo Bombers
Sep 12 2019 - The External Light Meter: Your New Best Friend
Sep 11 2019 - Bridal Gown: Capturing the Details of the Dress
Sep 10 2019 - Shooting a Wedding
Sep 09 2019 - The Growth of Baby Jill
Sep 08 2019 - Jake Neopolt, “Fire in the Sky” album cover
Sep 07 2019 - Chen Family
Sep 06 2019 - Juliane and Hanz
Sep 05 2019 - The Authentic Charm of Coney Island
Sep 04 2019 - Brooklyn Bridge – One Bridge, So Many Options


## Part V: Working with Files

### Writing

<p align="center"><em>The fundamental writing cheat-sheets can be found below.</em></p>


In this guide, the process of creating files in Python is explained using the `open()` function. The function is used to establish a connection to a file, and two parameters are provided: the file name and a mode flag that specifies how the file should be treated. There are three main mode options: 'x' for exclusive creation (only if the file doesn't exist), 'w' for writing (overwriting existing content or creating anew), and 'a' for appending (adding to the end of an existing file). The guide demonstrates the usage of these modes by creating a file named "cheese.txt" and writing various content to it. It emphasizes the importance of closing the file using the `close()` method to prevent issues. Additionally, the guide covers the use of the `sys.argv` list to accept command-line arguments for dynamic file naming. By following these steps, users can create and manipulate files in Python according to their specific needs.

#### Opening a File for Writing

##### Open a File in Write Mode
```python
file = open("filename.txt", "w")
```

##### Open a File in Append Mode
```python
file = open("filename.txt", "a")
```

#### Writing to a File

##### Write a String to the File
```python
file.write("Hello, World!")
```

##### Write Multiple Lines to the File
```python
lines = ["Line 1\n", "Line 2\n", "Line 3\n"]
file.writelines(lines)
```

##### Write with Automatic Line Break
```python
file.write("Line 1")
file.write("Line 2")
file.write("Line 3")
```

#### Closing a File

##### Close the File
```python
file.close()
```

#### Using the `with` Statement (Automatically Closes the File)

##### Write Using `with` Statement
```python
with open("filename.txt", "w") as file:
    file.write("Content")
```

#### Error Handling

##### Handle File Writing Errors
```python
try:
    with open("filename.txt", "w") as file:
        file.write("Content")
except IOError as e:
    print("An error occurred:", e)
```

#### File Flushing

##### Flush the Buffer to the File
```python
file.flush()
```

#### Writing Binary Data

##### Write Binary Data to a File
```python
with open("binary_file.bin", "wb") as file:
    data = b'\x00\x01\x02\x03'
    file.write(data)
```

#### Using `print()` Function to Write to a File

##### Write Using `print()` with a File Argument
```python
with open("filename.txt", "w") as file:
    print("Line 1", file=file)
    print("Line 2", file=file)
```

#### Note

Remember to replace `"filename.txt"` and `"binary_file.bin"` with your desired file names. Always handle exceptions appropriately and make sure to close files after writing to them.

### Reading

<p align="center"><em>The fundamental reading cheat-sheets can be found below.</em></p>

This tutorial outlines how to read an existing file in Python using the `open()` function. Readers are introduced to different techniques for reading file content. The guide starts by discussing the basics of file reading and the use of the `'r'` mode flag, which stands for read. The guide demonstrates reading the entire file content into a string using the `file.read()` method and shows how to read lines from the file into a list using `file.readlines()`. It also explains how to read the file line by line using a `for` loop. An example challenge is provided, encouraging readers to create a file with numbers and then develop a Python script to read and perform calculations on those numbers. The guide emphasizes the importance of properly closing files and offers practical insights into efficient file reading methods.

#### `open()` Function
```python
# Basic file reading
with open('filename.txt', 'r') as file:
    content = file.read()

## Read lines into a list
with open('filename.txt', 'r') as file:
    lines = file.readlines()

##### Iterating Through Lines
```python
with open('filename.txt', 'r') as file:
    for line in file:
        print(line)
```

##### Reading N Characters
```python
with open('filename.txt', 'r') as file:
    n_characters = 10
    content = file.read(n_characters)
```

##### Using `readline()`
```python
with open('filename.txt', 'r') as file:
    line1 = file.readline()
    line2 = file.readline()
```

#### Reading Binary Files

##### Reading Binary Data
```python
with open('file.bin', 'rb') as file:
    binary_data = file.read()
```

##### Reading Specific Number of Bytes
```python
with open('file.bin', 'rb') as file:
    bytes_to_read = 1024
    binary_data = file.read(bytes_to_read)
```

#### Using `os` Module

##### Checking File Existence
```python
import os

file_exists = os.path.exists('filename.txt')
```

##### Getting File Size
```python
file_size = os.path.getsize('filename.txt')
```

##### Checking if Path is a File
```python
is_file = os.path.isfile('filename.txt')
```

#### Using `pathlib` Module

##### Reading Text File
```python
from pathlib import Path

file_path = Path('filename.txt')
content = file_path.read_text()
```

##### Reading Binary File
```python
binary_file_path = Path('file.bin')
binary_data = binary_file_path.read_bytes()
```

Remember to replace `'filename.txt'` and `'file.bin'` with your actual file paths.

### Editing

<p align="center"><em>The fundamental editing cheat-sheets can be found below.</em></p>

When editing a file in Python, the process involves both reading an existing file and making changes to its content before saving the modified data. This combination of reading and writing is facilitated using the `open()` function. By providing a file path and an appropriate mode, such as 'r+' for both reading and writing, or 'w+' for reading and writing (which clears the file first), you can manipulate the file content.

To demonstrate a preferred method for file editing, a step-by-step approach is outlined. Firstly, the file is read using `file.read()` to retrieve its content and store it in a variable. Next, desired edits are made, such as appending, modifying lines, or inserting new lines. To implement these changes, the file is opened again, this time with 'w' mode, ensuring the previous content is overwritten. Within this stage, each line to be saved is written back using `file.write()`, which can accept a list of lines.

For instance, appending content involves reading the file, performing edits, and then writing the modified content back. Similarly, inserting new lines or modifying specific lines can be accomplished using index manipulation within the list of lines. It's essential to remember adding newline characters when writing lines to ensure proper formatting.

However, for adding content to the end of a file, a nuanced approach is required due to newline formatting peculiarities. To prevent issues, the last line's newline is retained, and a new line is appended before the additional content.

In conclusion, editing files in Python necessitates a strategy that combines reading, editing, and writing, with careful attention to formatting and newline characters. This method provides control over file modifications, ensuring efficient and accurate edits.

##### Writing to an Existing File
```python
with open('filename.txt', 'w') as file:
    file.write("Hello, world!\n")
```

##### Appending to a Text File
```python
with open('filename.txt', 'a') as file:
    file.write("This will be appended.\n")
```

#### Writing Binary Data

##### Writing Binary Data to a File
```python
binary_data = b'\x00\x01\x02'
with open('file.bin', 'wb') as file:
    file.write(binary_data)
```

#### Using `os` Module for File Operations

##### Renaming a File
```python
import os

os.rename('old_filename.txt', 'new_filename.txt')
```

##### Deleting a File
```python
import os

os.remove('filename.txt')
```

#### Using `pathlib` Module for File Operations

##### Renaming a File
```python
from pathlib import Path

old_path = Path('old_filename.txt')
new_path = Path('new_filename.txt')
old_path.rename(new_path)
```

##### Deleting a File
```python
from pathlib import Path

file_path = Path('filename.txt')
file_path.unlink()
```

#### Using `shutil` Module for File Operations

##### Copying a File
```python
import shutil

shutil.copy('source.txt', 'destination.txt')
```

##### Moving a File
```python
import shutil

shutil.move('old_location/file.txt', 'new_location/file.txt')
```

Remember to replace `'filename.txt'`, `'file.bin'`, `'old_filename.txt'`, `'new_filename.txt'`, `'source.txt'`, `'destination.txt'`, `'old_location/file.txt'`, and `'new_location/file.txt'` with your actual file paths.